### PR TITLE
Intrusive shared pointer for shamap inner nodes

### PR DESCRIPTION
**Pull request by seelabs**
I'm opening this draft PR so it can be tested for memory saving and transaction throughput. My own tests show a substantial memory savings (~20 - 25%, or ~4 gigabytes; see charts in the "draft_pr_remove_me" directory). I did not test transaction throughput.

This branch has a long history. About two years ago I wrote a patch to remove the mutex from shamap inner nodes (ref: https://github.com/seelabs/rippled/tree/lockfree-tagged-cache). At the time I measured a large memory savings of about 2 gig. Unfortunately, the code required using the `folly` library, and I was hesitant to introduce such a large dependency into rippled (especially one that was so hard to build). This branch resurrects that old work and removes the `folly` dependency.

The old branch used a lockless atomic shared pointer. This new branch introduces a intrusive pointer type. Unlike boost's intrusive pointer, this intrusive pointer can handle both strong and weak pointers (needed for the tagged cache). Since this is an intrusive pointer type, in order to support weak pointers, the object is not destroyed when the strong count goes to zero. Instead, it is "partially destroyed" (for example, inner nodes will reset their children). This intrusive pointer takes 32-bits for both the strong and weak counts, and takes one 64-bit pointer to point at the object. This is much smaller than a std::shared_pointer, which needs a control block to hold the strong and weak counts (and potentially other objects), as well as an extra pointer to point at the control block. The intrusive shared pointer also has optional support for atomic operations. These atomic operations can be used instead of the lock when changing inner node pointers in the shamap.

Note: The space savings is independent from removing the locks from shamap inner node. Even if we keep the lock we may want to keep this patch for the space savings (and it will take careful auditing to confirm we can remove the locks).

Note: Intrusive pointer will be 12 bytes. The shared_ptr will be around 40 bytes, depending on implementation.

### CODE REVIEW

**Comment by sublimator** [comment]
Collapse the single inner node child inners (to the books base) while
maintaining hash repro and back-compat?

**Comment by seelabs** [comment]
@sublimator Hey! Good to see you! No, this is different. The shamap keeps it's nodes as `std::shared_ptr` and the tagged caches also keep tree node as `std::shared_ptr`. A `shared_ptr` takes about 40 bytes (depending on implementation). The intrusive pointer in this patch takes about 12 bytes. That matters because rippled keeps lots and lots of tree nodes around. The other problem with `std::shared_ptr` is you can't change it atomically. The intrusive pointer in this patch has support for atomic updates (which can be used to remove the inner node locks, I _think_).

**Comment by sublimator** [comment]
Hi :) I was hoping you would be all "challenge accepted!" lol

**Comment by sublimator** [comment]
Fun aside - that sounds good

**Comment by sublimator** [comment]
Actually:

```{depth-0=1, depth-1=16, depth-2=256, depth-3=4096, depth-4=62961,
depth-5=42846, depth-6=4107, depth-7=1204, depth-8=1004, depth-9=986,
depth-10=986, depth-11=986, depth-12=986, depth-13=986, depth-14=986,
depth-15=986, depth-16=986, depth-17=986, depth-18=986, depth-19=986,
depth-20=986, depth-21=986, depth-22=986, depth-23=986, depth-24=986,
depth-25=986, depth-26=986, depth-27=986, depth-28=986, depth-29=986,
depth-30=986, depth-31=986, depth-32=986, depth-33=986, depth-34=986,
depth-35=986, depth-36=986, depth-37=986, depth-38=986, depth-39=986,
depth-40=986, depth-41=986, depth-42=986, depth-43=986, depth-44=986,
depth-45=986, depth-46=986, depth-47=986, depth-48=986, depth-49=1040,
depth-50=1214, depth-51=1399, depth-52=1974, depth-53=2118, depth-54=1049,
depth-55=553, depth-56=392, depth-57=333, depth-58=283, depth-59=225,
depth-60=148, depth-61=91, depth-62=75, depth-63=71, totalSlots=2670336,
emptySlots=2159209, singleInnerChild=49955, totalInners=166895,
totalLeaves=344232} 
```

Collapsed and maintaining hash (lazy synthetic skipped inners) repro:

```
{depth-0=1, depth-1=16, depth-2=256, depth-3=4096, depth-4=62961,
depth-5=42497, depth-6=1782, depth-7=990, depth-8=1135, depth-9=1278,
depth-10=1164, depth-11=603, depth-12=135, depth-13=25, depth-14=2,
totalSlots=1871056, emptySlots=1409884, singleInnerChild=0, totalInners=116940,
totalLeaves=344232}
```

166895 vs 116940 inners for 344232 leaves 

Tricky with NodeStore integration/migration headaches but intrusive pointers + collapsed inners combined could save a lot of bytes?

New node serialization (branch header + non empty branches + etc ) not all that different to already used snappy conceptually ? Just another compression.

**Comment by sublimator** [comment]
Pulled a history pack out of the glacier and ran some numbers, which I believe speak for themselves. Even if only for new chains it might be worth looking into ShaMap v2 again? 

| ledger | effectedOffers | txns | newInners |
|--------|----------------|------|-----------|
| [11375001](https://xrpscan.com/ledger/11375001) | 40 | 45 | unknown |
| [11375002](https://xrpscan.com/ledger/11375002) | 65 | 49 | 1570 |
| [11375003](https://xrpscan.com/ledger/11375003) | 50 | 41 | 1267 |
| [11375004](https://xrpscan.com/ledger/11375004) | 25 | 21 | 783 |
| [11375005](https://xrpscan.com/ledger/11375005) | 43 | 34 | 978 |
| [11375006](https://xrpscan.com/ledger/11375006) | 41 | 28 | 1052 |
| [11375007](https://xrpscan.com/ledger/11375007) | 45 | 41 | 1126 |
| [11375008](https://xrpscan.com/ledger/11375008) | 58 | 50 | 1443 |
| [11375009](https://xrpscan.com/ledger/11375009) | 32 | 28 | 1079 |
| [11375010](https://xrpscan.com/ledger/11375010) | 55 | 51 | 1547 |
| [11375011](https://xrpscan.com/ledger/11375011) | 39 | 65 | 1267 |
| [11375012](https://xrpscan.com/ledger/11375012) | 39 | 44 | 1166 |
| [11375013](https://xrpscan.com/ledger/11375013) | 24 | 24 | 756 |
| [11375014](https://xrpscan.com/ledger/11375014) | 17 | 13 | 553 |
| [11375015](https://xrpscan.com/ledger/11375015) | 44 | 42 | 1232 |
| [11375016](https://xrpscan.com/ledger/11375016) | 24 | 26 | 702 |
| [11375017](https://xrpscan.com/ledger/11375017) | 22 | 34 | 788 |
| [11375018](https://xrpscan.com/ledger/11375018) | 16 | 15 | 508 |
| [11375019](https://xrpscan.com/ledger/11375019) | 17 | 23 | 580 |
| [11375020](https://xrpscan.com/ledger/11375020) | 59 | 57 | 1795 |
| [11375021](https://xrpscan.com/ledger/11375021) | 47 | 46 | 1313 |
| [11375022](https://xrpscan.com/ledger/11375022) | 60 | 59 | 1831 |
| [11375023](https://xrpscan.com/ledger/11375023) | 49 | 45 | 1376 |
| [11375024](https://xrpscan.com/ledger/11375024) | 77 | 77 | 1990 |
| [11375025](https://xrpscan.com/ledger/11375025) | 43 | 35 | 1238 |
| [11375026](https://xrpscan.com/ledger/11375026) | 35 | 35 | 912 |
| [11375027](https://xrpscan.com/ledger/11375027) | 20 | 28 | 690 |
| [11375028](https://xrpscan.com/ledger/11375028) | 34 | 38 | 1009 |
| [11375029](https://xrpscan.com/ledger/11375029) | 12 | 11 | 514 |
| [11375030](https://xrpscan.com/ledger/11375030) | 15 | 15 | 518 |
| [11375031](https://xrpscan.com/ledger/11375031) | 8 | 9 | 372 |
| [11375032](https://xrpscan.com/ledger/11375032) | 6 | 7 | 252 |
| [11375033](https://xrpscan.com/ledger/11375033) | 36 | 39 | 1303 |
| [11375034](https://xrpscan.com/ledger/11375034) | 27 | 23 | 753 |
| [11375035](https://xrpscan.com/ledger/11375035) | 38 | 32 | 889 |
| [11375036](https://xrpscan.com/ledger/11375036) | 23 | 16 | 722 |
| [11375037](https://xrpscan.com/ledger/11375037) | 43 | 27 | 1098 |
| [11375038](https://xrpscan.com/ledger/11375038) | 39 | 51 | 965 |
| [11375039](https://xrpscan.com/ledger/11375039) | 117 | 91 | 2578 |
| [11375040](https://xrpscan.com/ledger/11375040) | 63 | 62 | 1475 |
| [11375041](https://xrpscan.com/ledger/11375041) | 48 | 53 | 1172 |
| [11375042](https://xrpscan.com/ledger/11375042) | 43 | 44 | 1182 |
| [11375043](https://xrpscan.com/ledger/11375043) | 57 | 53 | 1361 |
| [11375044](https://xrpscan.com/ledger/11375044) | 15 | 13 | 565 |
| [11375045](https://xrpscan.com/ledger/11375045) | 14 | 13 | 508 |
| [11375046](https://xrpscan.com/ledger/11375046) | 12 | 14 | 501 |
| [11375047](https://xrpscan.com/ledger/11375047) | 12 | 13 | 425 |
| [11375048](https://xrpscan.com/ledger/11375048) | 19 | 25 | 711 |
| [11375049](https://xrpscan.com/ledger/11375049) | 41 | 40 | 1079 |
| [11375050](https://xrpscan.com/ledger/11375050) | 32 | 33 | 973 |
| [11375051](https://xrpscan.com/ledger/11375051) | 26 | 21 | 677 |
| [11375052](https://xrpscan.com/ledger/11375052) | 15 | 21 | 382 |
| [11375053](https://xrpscan.com/ledger/11375053) | 15 | 18 | 740 |
| [11375054](https://xrpscan.com/ledger/11375054) | 39 | 29 | 995 |
| [11375055](https://xrpscan.com/ledger/11375055) | 72 | 63 | 1277 |
| [11375056](https://xrpscan.com/ledger/11375056) | 30 | 31 | 845 |
| [11375057](https://xrpscan.com/ledger/11375057) | 39 | 35 | 1083 |
| [11375058](https://xrpscan.com/ledger/11375058) | 76 | 70 | 2054 |
| [11375059](https://xrpscan.com/ledger/11375059) | 63 | 127 | 1809 |
| [11375060](https://xrpscan.com/ledger/11375060) | 66 | 68 | 1837 |
| [11375061](https://xrpscan.com/ledger/11375061) | 64 | 58 | 1529 |
| [11375062](https://xrpscan.com/ledger/11375062) | 36 | 38 | 913 |
| [11375063](https://xrpscan.com/ledger/11375063) | 48 | 44 | 1225 |
| [11375064](https://xrpscan.com/ledger/11375064) | 55 | 50 | 1338 |
| [11375065](https://xrpscan.com/ledger/11375065) | 32 | 28 | 769 |
| [11375066](https://xrpscan.com/ledger/11375066) | 20 | 17 | 665 |
| [11375067](https://xrpscan.com/ledger/11375067) | 51 | 40 | 1156 |
| [11375068](https://xrpscan.com/ledger/11375068) | 22 | 17 | 628 |
| [11375069](https://xrpscan.com/ledger/11375069) | 49 | 31 | 1020 |
| [11375070](https://xrpscan.com/ledger/11375070) | 72 | 42 | 1682 |
| [11375071](https://xrpscan.com/ledger/11375071) | 40 | 31 | 988 |
| [11375072](https://xrpscan.com/ledger/11375072) | 44 | 33 | 1021 |
| [11375073](https://xrpscan.com/ledger/11375073) | 47 | 40 | 1012 |
| [11375074](https://xrpscan.com/ledger/11375074) | 25 | 22 | 771 |
| [11375075](https://xrpscan.com/ledger/11375075) | 19 | 15 | 560 |
| [11375076](https://xrpscan.com/ledger/11375076) | 21 | 14 | 567 |
| [11375077](https://xrpscan.com/ledger/11375077) | 31 | 18 | 862 |
| [11375078](https://xrpscan.com/ledger/11375078) | 24 | 18 | 702 |
| [11375079](https://xrpscan.com/ledger/11375079) | 38 | 29 | 1197 |
| [11375080](https://xrpscan.com/ledger/11375080) | 38 | 33 | 1133 |
| [11375081](https://xrpscan.com/ledger/11375081) | 37 | 33 | 1104 |
| [11375082](https://xrpscan.com/ledger/11375082) | 39 | 40 | 1137 |
| [11375083](https://xrpscan.com/ledger/11375083) | 28 | 25 | 1086 |
| [11375084](https://xrpscan.com/ledger/11375084) | 35 | 29 | 1062 |
| [11375085](https://xrpscan.com/ledger/11375085) | 56 | 46 | 1390 |
| [11375086](https://xrpscan.com/ledger/11375086) | 43 | 31 | 1358 |
| [11375087](https://xrpscan.com/ledger/11375087) | 43 | 43 | 1121 |
| [11375088](https://xrpscan.com/ledger/11375088) | 52 | 52 | 1392 |
| [11375089](https://xrpscan.com/ledger/11375089) | 82 | 63 | 3173 |
| [11375090](https://xrpscan.com/ledger/11375090) | 22 | 17 | 813 |
| [11375091](https://xrpscan.com/ledger/11375091) | 19 | 25 | 837 |
| [11375092](https://xrpscan.com/ledger/11375092) | 30 | 27 | 1041 |
| [11375093](https://xrpscan.com/ledger/11375093) | 41 | 37 | 1490 |
| [11375094](https://xrpscan.com/ledger/11375094) | 38 | 38 | 1222 |
| [11375095](https://xrpscan.com/ledger/11375095) | 31 | 28 | 1014 |
| [11375096](https://xrpscan.com/ledger/11375096) | 34 | 31 | 988 |
| [11375097](https://xrpscan.com/ledger/11375097) | 44 | 41 | 1207 |
| [11375098](https://xrpscan.com/ledger/11375098) | 33 | 31 | 1145 |
| [11375099](https://xrpscan.com/ledger/11375099) | 37 | 29 | 1169 |
| [11375100](https://xrpscan.com/ledger/11375100) | 28 | 28 | 927 |
| [11375101](https://xrpscan.com/ledger/11375101) | 48 | 35 | 1360 |
| [11375102](https://xrpscan.com/ledger/11375102) | 29 | 32 | 1153 |
| [11375103](https://xrpscan.com/ledger/11375103) | 39 | 37 | 1301 |
| [11375104](https://xrpscan.com/ledger/11375104) | 13 | 13 | 563 |
| [11375105](https://xrpscan.com/ledger/11375105) | 18 | 19 | 656 |
| [11375106](https://xrpscan.com/ledger/11375106) | 53 | 54 | 1240 |
| [11375107](https://xrpscan.com/ledger/11375107) | 26 | 26 | 954 |
| [11375108](https://xrpscan.com/ledger/11375108) | 32 | 26 | 1209 |
| [11375109](https://xrpscan.com/ledger/11375109) | 21 | 24 | 922 |
| [11375110](https://xrpscan.com/ledger/11375110) | 13 | 18 | 585 |
| [11375111](https://xrpscan.com/ledger/11375111) | 17 | 14 | 663 |
| [11375112](https://xrpscan.com/ledger/11375112) | 18 | 22 | 829 |
| [11375113](https://xrpscan.com/ledger/11375113) | 36 | 35 | 1179 |
| [11375114](https://xrpscan.com/ledger/11375114) | 23 | 25 | 918 |
| [11375115](https://xrpscan.com/ledger/11375115) | 11 | 16 | 542 |
| [11375116](https://xrpscan.com/ledger/11375116) | 12 | 13 | 645 |
| [11375117](https://xrpscan.com/ledger/11375117) | 137 | 105 | 2971 |
| [11375118](https://xrpscan.com/ledger/11375118) | 126 | 123 | 2833 |

**Comment by sublimator** [comment]
With inners / book bases:

| ledger | effectedOffers | txns | book bases | newInners  | ratio |
|--------|----------------|------|------------|------------|--------
| [11375001](https://xrpscan.com/ledger/11375001) | 40 | 45 | 19 | unknown | unknown | 
| [11375002](https://xrpscan.com/ledger/11375002) | 65 | 49 | 24 | 1570 | 65 |
| [11375003](https://xrpscan.com/ledger/11375003) | 50 | 41 | 20 | 1267 | 63 |
| [11375004](https://xrpscan.com/ledger/11375004) | 25 | 21 | 12 | 783 | 65 |
| [11375005](https://xrpscan.com/ledger/11375005) | 43 | 34 | 14 | 978 | 69 |
| [11375006](https://xrpscan.com/ledger/11375006) | 41 | 28 | 17 | 1052 | 61 |
| [11375007](https://xrpscan.com/ledger/11375007) | 45 | 41 | 18 | 1126 | 62 |
| [11375008](https://xrpscan.com/ledger/11375008) | 58 | 50 | 23 | 1443 | 62 |
| [11375009](https://xrpscan.com/ledger/11375009) | 32 | 28 | 17 | 1079 | 63 |
| [11375010](https://xrpscan.com/ledger/11375010) | 55 | 51 | 25 | 1547 | 61 |
| [11375011](https://xrpscan.com/ledger/11375011) | 39 | 65 | 23 | 1267 | 55 |
| [11375012](https://xrpscan.com/ledger/11375012) | 39 | 44 | 17 | 1166 | 68 |
| [11375013](https://xrpscan.com/ledger/11375013) | 24 | 24 | 13 | 756 | 58 |
| [11375014](https://xrpscan.com/ledger/11375014) | 17 | 13 | 8 | 553 | 69 |
| [11375015](https://xrpscan.com/ledger/11375015) | 44 | 42 | 20 | 1232 | 61 |
| [11375016](https://xrpscan.com/ledger/11375016) | 24 | 26 | 11 | 702 | 63 |
| [11375017](https://xrpscan.com/ledger/11375017) | 22 | 34 | 14 | 788 | 56 |
| [11375018](https://xrpscan.com/ledger/11375018) | 16 | 15 | 8 | 508 | 63 |
| [11375019](https://xrpscan.com/ledger/11375019) | 17 | 23 | 9 | 580 | 64 |
| [11375020](https://xrpscan.com/ledger/11375020) | 59 | 57 | 28 | 1795 | 64 |
| [11375021](https://xrpscan.com/ledger/11375021) | 47 | 46 | 23 | 1313 | 57 |
| [11375022](https://xrpscan.com/ledger/11375022) | 60 | 59 | 33 | 1831 | 55 |
| [11375023](https://xrpscan.com/ledger/11375023) | 49 | 45 | 21 | 1376 | 65 |
| [11375024](https://xrpscan.com/ledger/11375024) | 77 | 77 | 33 | 1990 | 60 |
| [11375025](https://xrpscan.com/ledger/11375025) | 43 | 35 | 24 | 1238 | 51 |
| [11375026](https://xrpscan.com/ledger/11375026) | 35 | 35 | 14 | 912 | 65 |
| [11375027](https://xrpscan.com/ledger/11375027) | 20 | 28 | 11 | 690 | 62 |
| [11375028](https://xrpscan.com/ledger/11375028) | 34 | 38 | 16 | 1009 | 63 |
| [11375029](https://xrpscan.com/ledger/11375029) | 12 | 11 | 8 | 514 | 64 |
| [11375030](https://xrpscan.com/ledger/11375030) | 15 | 15 | 9 | 518 | 57 |
| [11375031](https://xrpscan.com/ledger/11375031) | 8 | 9 | 6 | 372 | 62 |
| [11375032](https://xrpscan.com/ledger/11375032) | 6 | 7 | 4 | 252 | 63 |
| [11375033](https://xrpscan.com/ledger/11375033) | 36 | 39 | 22 | 1303 | 59 |
| [11375034](https://xrpscan.com/ledger/11375034) | 27 | 23 | 15 | 753 | 50 |
| [11375035](https://xrpscan.com/ledger/11375035) | 38 | 32 | 19 | 889 | 46 |
| [11375036](https://xrpscan.com/ledger/11375036) | 23 | 16 | 12 | 722 | 60 |
| [11375037](https://xrpscan.com/ledger/11375037) | 43 | 27 | 17 | 1098 | 64 |
| [11375038](https://xrpscan.com/ledger/11375038) | 39 | 51 | 19 | 965 | 50 |
| [11375039](https://xrpscan.com/ledger/11375039) | 117 | 91 | 40 | 2578 | 64 |
| [11375040](https://xrpscan.com/ledger/11375040) | 63 | 62 | 25 | 1475 | 59 |
| [11375041](https://xrpscan.com/ledger/11375041) | 48 | 53 | 19 | 1172 | 61 |
| [11375042](https://xrpscan.com/ledger/11375042) | 43 | 44 | 19 | 1182 | 62 |
| [11375043](https://xrpscan.com/ledger/11375043) | 57 | 53 | 20 | 1361 | 68 |
| [11375044](https://xrpscan.com/ledger/11375044) | 15 | 13 | 10 | 565 | 56 |
| [11375045](https://xrpscan.com/ledger/11375045) | 14 | 13 | 7 | 508 | 72 |
| [11375046](https://xrpscan.com/ledger/11375046) | 12 | 14 | 8 | 501 | 62 |
| [11375047](https://xrpscan.com/ledger/11375047) | 12 | 13 | 6 | 425 | 70 |
| [11375048](https://xrpscan.com/ledger/11375048) | 19 | 25 | 12 | 711 | 59 |
| [11375049](https://xrpscan.com/ledger/11375049) | 41 | 40 | 18 | 1079 | 59 |
| [11375050](https://xrpscan.com/ledger/11375050) | 32 | 33 | 17 | 973 | 57 |
| [11375051](https://xrpscan.com/ledger/11375051) | 26 | 21 | 10 | 677 | 67 |
| [11375052](https://xrpscan.com/ledger/11375052) | 15 | 21 | 5 | 382 | 76 |
| [11375053](https://xrpscan.com/ledger/11375053) | 15 | 18 | 11 | 740 | 67 |
| [11375054](https://xrpscan.com/ledger/11375054) | 39 | 29 | 16 | 995 | 62 |
| [11375055](https://xrpscan.com/ledger/11375055) | 72 | 63 | 22 | 1277 | 58 |
| [11375056](https://xrpscan.com/ledger/11375056) | 30 | 31 | 14 | 845 | 60 |
| [11375057](https://xrpscan.com/ledger/11375057) | 39 | 35 | 18 | 1083 | 60 |
| [11375058](https://xrpscan.com/ledger/11375058) | 76 | 70 | 36 | 2054 | 57 |
| [11375059](https://xrpscan.com/ledger/11375059) | 63 | 127 | 28 | 1809 | 64 |
| [11375060](https://xrpscan.com/ledger/11375060) | 66 | 68 | 32 | 1837 | 57 |
| [11375061](https://xrpscan.com/ledger/11375061) | 64 | 58 | 28 | 1529 | 54 |
| [11375062](https://xrpscan.com/ledger/11375062) | 36 | 38 | 14 | 913 | 65 |
| [11375063](https://xrpscan.com/ledger/11375063) | 48 | 44 | 19 | 1225 | 64 |
| [11375064](https://xrpscan.com/ledger/11375064) | 55 | 50 | 20 | 1338 | 66 |
| [11375065](https://xrpscan.com/ledger/11375065) | 32 | 28 | 11 | 769 | 69 |
| [11375066](https://xrpscan.com/ledger/11375066) | 20 | 17 | 10 | 665 | 66 |
| [11375067](https://xrpscan.com/ledger/11375067) | 51 | 40 | 16 | 1156 | 72 |
| [11375068](https://xrpscan.com/ledger/11375068) | 22 | 17 | 9 | 628 | 69 |
| [11375069](https://xrpscan.com/ledger/11375069) | 49 | 31 | 14 | 1020 | 72 |
| [11375070](https://xrpscan.com/ledger/11375070) | 72 | 42 | 26 | 1682 | 64 |
| [11375071](https://xrpscan.com/ledger/11375071) | 40 | 31 | 14 | 988 | 70 |
| [11375072](https://xrpscan.com/ledger/11375072) | 44 | 33 | 15 | 1021 | 68 |
| [11375073](https://xrpscan.com/ledger/11375073) | 47 | 40 | 15 | 1012 | 67 |
| [11375074](https://xrpscan.com/ledger/11375074) | 25 | 22 | 12 | 771 | 64 |
| [11375075](https://xrpscan.com/ledger/11375075) | 19 | 15 | 8 | 560 | 70 |
| [11375076](https://xrpscan.com/ledger/11375076) | 21 | 14 | 8 | 567 | 70 |
| [11375077](https://xrpscan.com/ledger/11375077) | 31 | 18 | 13 | 862 | 66 |
| [11375078](https://xrpscan.com/ledger/11375078) | 24 | 18 | 10 | 702 | 70 |
| [11375079](https://xrpscan.com/ledger/11375079) | 38 | 29 | 18 | 1197 | 66 |
| [11375080](https://xrpscan.com/ledger/11375080) | 38 | 33 | 21 | 1133 | 53 |
| [11375081](https://xrpscan.com/ledger/11375081) | 37 | 33 | 18 | 1104 | 61 |
| [11375082](https://xrpscan.com/ledger/11375082) | 39 | 40 | 20 | 1137 | 56 |
| [11375083](https://xrpscan.com/ledger/11375083) | 28 | 25 | 17 | 1086 | 63 |
| [11375084](https://xrpscan.com/ledger/11375084) | 35 | 29 | 19 | 1062 | 55 |
| [11375085](https://xrpscan.com/ledger/11375085) | 56 | 46 | 21 | 1390 | 66 |
| [11375086](https://xrpscan.com/ledger/11375086) | 43 | 31 | 21 | 1358 | 64 |
| [11375087](https://xrpscan.com/ledger/11375087) | 43 | 43 | 17 | 1121 | 65 |
| [11375088](https://xrpscan.com/ledger/11375088) | 52 | 52 | 24 | 1392 | 58 |
| [11375089](https://xrpscan.com/ledger/11375089) | 82 | 63 | 61 | 3173 | 52 |
| [11375090](https://xrpscan.com/ledger/11375090) | 22 | 17 | 13 | 813 | 62 |
| [11375091](https://xrpscan.com/ledger/11375091) | 19 | 25 | 13 | 837 | 64 |
| [11375092](https://xrpscan.com/ledger/11375092) | 30 | 27 | 16 | 1041 | 65 |
| [11375093](https://xrpscan.com/ledger/11375093) | 41 | 37 | 23 | 1490 | 64 |
| [11375094](https://xrpscan.com/ledger/11375094) | 38 | 38 | 19 | 1222 | 64 |
| [11375095](https://xrpscan.com/ledger/11375095) | 31 | 28 | 15 | 1014 | 67 |
| [11375096](https://xrpscan.com/ledger/11375096) | 34 | 31 | 17 | 988 | 58 |
| [11375097](https://xrpscan.com/ledger/11375097) | 44 | 41 | 19 | 1207 | 63 |
| [11375098](https://xrpscan.com/ledger/11375098) | 33 | 31 | 19 | 1145 | 60 |
| [11375099](https://xrpscan.com/ledger/11375099) | 37 | 29 | 18 | 1169 | 64 |
| [11375100](https://xrpscan.com/ledger/11375100) | 28 | 28 | 14 | 927 | 66 |
| [11375101](https://xrpscan.com/ledger/11375101) | 48 | 35 | 21 | 1360 | 64 |

**Comment by sublimator** [comment]
@seelabs 

Hi, I tried to build this branch as I'm curious about the perf boost, but it seems even with the latest XCode installed, std::atomic_ref isn't available in my SDKs. I am not sure I'm not doing something silly though. 

You're on Linux?

**Comment by seelabs** [comment]
@sublimator Yeah, I'm on linux. I'm disappointed the latest xcode doesn't have `atomic_ref`. That's not essential, but it would have been nice. I can take it out.

**Comment by seelabs** [comment]
I'm closing this PR (it was a draft anyway). There are some serious threading issues I need to think through.

**Comment by sublimator** [comment]
> I'm disappointed the latest xcode doesn't have atomic_ref.

I might have messed something up, but it really seemed that way

File: src/ripple/shamap/impl/SHAMapInnerNode.cpp [hunk 1]
```diff
@@ -326,7 +327,8 @@ SHAMapInnerNode::setChild(int m, intr_ptr::SharedPtr<SHAMapTreeNode> child)
         auto const childIndex = *getChildIndex(m);
         auto [_, hashes, children] = hashesAndChildren_.getHashesAndChildren();
         hashes[childIndex].zero();
-        children[childIndex] = std::move(child);
+        children[childIndex].assign(
```
Comment (sublimator) [inline src/ripple/shamap/impl/SHAMapInnerNode.cpp:330]: I'm still banging my head against the rippled SHAMap, as I have little experience with their "top-down" use for syncing - another world of headache.

If you only enter from the top, seems you're gold with CoW, but it seems like rippled needs to modify a tree while it's in use for syncing purposes? So you need kind of write-only nodes for a time? In rippled are the syncing threads partitioned to certain keyspaces?

**Pull request closed without merging**